scripts/opportunity-provider/index.ts

```diff
@@ -32,43 +36,69 @@ interface Config {
   rpcUrl: string;
 }
 
-function readConfig(path: string): Config {
-  const data = fs.readFileSync(path, "utf8");
-  return JSON.parse(data) as Config;
+const decimals: Record<Address, number> = {};
+const prices: Record<string, number> = {};
+
+async function getDecimals(config: Config, token: Token): Promise<number> {
+  const index = token.address;
+  if (decimals[index]) {
+    return decimals[index];
+  }
+
+  const client = getClient(config);
+  decimals[index] = await client.readContract({
+    address: token.address,
+    abi: erc20Abi,
+    functionName: "decimals",
+  });
+  return decimals[index];
 }
 
-function readOpportunity(path: string): Opportunity {
+function readFile<T>(path: string): T {
   const data = fs.readFileSync(path, "utf8");
-  JSON.parse(data) as Opportunity;
-  return JSON.parse(data) as Opportunity;
+  return JSON.parse(data) as T;
 }
 
-function getClinet(config: Config) {
-  return createPublicClient({
+let client: PublicClient | undefined = undefined;
```

Comment:
this caching seems not necessary. We are not spending any calls on creating the client